### PR TITLE
feat: [CDS-34522]: Add  TemplateLinkConfig in PipelineInforCondif DTO

### PR DESCRIPTION
**Pull request by abhinavmittal-harness**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/31177)

### CODE REVIEW

**Review by inderschera — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31177)* status: 2 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[860-orchestration-steps/src/main/java/io/harness/plancreator/pipeline/PipelineInfoConfig.java, line 79 at r1](https://reviewable.io/reviews/harness/harness-core/31177#-MyBPf-l78XiC6-wUUag:-MyBPf-l78XiC6-wUUah:bdn7zx3) ([raw file](https://github.com/harness/harness-core/blob/97660f31ffa4bc693cc203a5442ab4b5e792c5fe/860-orchestration-steps/src/main/java/io/harness/plancreator/pipeline/PipelineInfoConfig.java#L79)):*
> ```Java
>   ParameterField<Timeout> timeout;
> 
>   TemplateLinkConfig template;
> ```

This will start showing up in auto suggestion in pipeline yaml along with stages and timeout, which is not expected.
Once we have written template in pipeline yaml, stages should not come in auto suggest.
We should add @OneOfSet on this to avoid that.
For reference, check StageElementConfig. 
We should test this part out in yaml view in harness UI.

File: 840-template-service/src/main/java/io/harness/template/handler/PipelineTemplateYamlConversionHandler.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      TemplateEntityType templateEntityType, YamlField yamlField) {
+    Map<String, Object> fieldsToAdd = new HashMap<>();
+    fieldsToAdd.put(IDENTIFIER, TEMPLATE_INPUTS);
+    fieldsToAdd.put(NAME, TEMPLATE_INPUTS);
+    String rootYamlFieldName = getRootField(templateEntityType);
+    YamlField rootYamlField = yamlField.getNode().getField(rootYamlFieldName);
+    if (rootYamlField == null) {
+      throw new NGTemplateException("yamlNode provided doesn not have root yaml field: " + rootYamlFieldName);
+    }
+    YamlField typeYamlField = rootYamlField.getNode().getField(STAGES);
```
Comment (abhinavmittal-harness) [inline 840-template-service/src/main/java/io/harness/template/handler/PipelineTemplateYamlConversionHandler.java:43]: I'll fix the variable name in the subsequent commit

**Review by inderschera — COMMENTED** [review]
> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31177)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[840-template-service/src/main/java/io/harness/template/handler/PipelineTemplateYamlConversionHandler.java, line 45 at r2](https://reviewable.io/reviews/harness/harness-core/31177#-MzEe1SB6o6SFvsSbaM3:-MzEe1SB6o6SFvsSbaM4:boxu98c) ([raw file](https://github.com/harness/harness-core/blob/da47f08540deac072203bddf9b08d1d4a61c432b/840-template-service/src/main/java/io/harness/template/handler/PipelineTemplateYamlConversionHandler.java#L45)):*
> ```Java
>     YamlField typeYamlField = rootYamlField.getNode().getField(STAGES);
>     if (typeYamlField == null) {
>       throw new NGTemplateException("yamlNode provided doesn not have type yaml field");
> ```

Please fix type -> stages here as well.

**Review by inderschera — DISMISSED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31177#-:-MzEo65hENMcm46qkzD9:bnfp4nl)* status: :shipit: complete! ✓

**Review by poojaSinghal — APPROVED** [review]
> Reviewed 5 of 7 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31177)* status: :shipit: complete! ✓

**Review by inderschera — COMMENTED** [review]
> Reviewed 1 of 2 files at r3, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31177)* status: 0 of 1 approvals obtained.

**Review by inderschera — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31177)* status: 0 of 1 approvals obtained.

**Review by inderschera — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31177)* status: :shipit: complete! ✓

**Pull request merged**